src/main/java/build/buildfarm/worker/OutputDirectory.java

```diff
@@ -0,0 +1,82 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Stack;
+
+// would be great to be able to inherit from ImmutableMap...
+public class OutputDirectory extends HashMap<String, OutputDirectory> {
+  private OutputDirectory() {
+  }
+
+  public static OutputDirectory parse(Iterable<String> outputFiles, Iterable<String> outputDirs) {
+    return parseDirectories(Iterables.mergeSorted(
+        ImmutableList.<Iterable<String>>of(
+            Iterables.transform(
+                outputFiles,
+                (file) -> file.contains("/") ? file.substring(0, file.lastIndexOf('/') + 1) : ""),
+            outputDirs),
+        (a, b) -> a.compareTo(b)));
+  }
+
+  private static OutputDirectory parseDirectories(Iterable<String> outputDirs) {
```

Comment:
Do I understand correctly that the only usage of this class is so you could create all the directories involved (stamp)?
If that is the case, why not use mkdirs every time you need to create an output, like https://github.com/bazelbuild/bazel/blob/master/src/tools/remote/src/main/java/com/google/devtools/build/remote/worker/ExecutionServer.java#L201 ?

Reply:
It is actually used as an efficient predicate for the symlinked directory optimization - a `linkInputs` subdirectory cannot be symlinked if it leads to an output path. Output files must have real and operation independent directories all the way up to the `execRoot`.

Reply:
Ah, thank you. In that case, a nit: I think composition is better than inheritance. You want expose the users of this class to a very narrow subset of the HashMap interface (no modification, iteration not necessary either, just is-in check), and in fact it's not even a subset, the fact that you use a hashmap and not something else is an implementation detail. For me, a reason to inherit from X is only if I need to use instances of MyClass outside of MyClass in contexts where an X is expected.

Reply:
Completely agree with that, I'll cut it down.